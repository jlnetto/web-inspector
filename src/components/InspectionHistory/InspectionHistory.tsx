import { useState, useEffect } from 'react';
import type { StoredData } from '../../types/core';
import {
  AccordionContainer,
  AccordionContent,
  AccordionHeader,
  AccordionItem,
  ChevronIcon,
  Container,
  EmptyStateContainer,
  EmptyStateContent,
  Title,
} from './styles';
import InspectionDetails from '../InspectionDetails/InspectionDetails';

const InspectionHistory = () => {
  const [lastKeywords, setLastKeywords] = useState<StoredData[]>([]);
  const [expandedItemId, setExpandedItemId] = useState<string>('');

  const toggleAccordion = (itemId: string) => {
    setExpandedItemId((prevExpandedItemId) =>
      prevExpandedItemId === itemId ? '' : itemId
    );
  };

  useEffect(() => {
    const storedKeywords = localStorage.getItem('lastKeywords');
    if (storedKeywords) {
      setLastKeywords(JSON.parse(storedKeywords));
    }

    const handleStorageChange = (event: StorageEvent) => {
      if (event.key === 'lastKeywords') {
        const updatedKeywords = event.newValue
          ? JSON.parse(event.newValue)
          : [];
        setLastKeywords(updatedKeywords);
      }
    };

    window.addEventListener('storage', handleStorageChange);

    return () => {
      window.removeEventListener('storage', handleStorageChange);
    };
  }, []);

  return (
    <Container>
      <Title>Histórico de solicitações cadastradas</Title>
      {lastKeywords.length > 0 ? (
        <AccordionContainer>
          {lastKeywords.map(({ id, keyword }: StoredData) => (
            <AccordionItem key={id}>
              <AccordionHeader onClick={() => toggleAccordion(id)}>
                {keyword}
                <ChevronIcon expanded={expandedItemId === id} />
              </AccordionHeader>
              <AccordionContent expanded={expandedItemId === id}>
                <InspectionDetails id={id} />
              </AccordionContent>
            </AccordionItem>
          ))}
        </AccordionContainer>
      ) : (
        <EmptyStateContainer>
          <EmptyStateContent>No records</EmptyStateContent>
        </EmptyStateContainer>
      )}
    </Container>
  );
};

export default InspectionHistory;
