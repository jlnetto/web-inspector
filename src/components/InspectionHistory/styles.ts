import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;
`;

export const Title = styled.h2`
  font-size: 20px;
  margin: 0;
`;

export const AccordionContainer = styled.div``;

export const AccordionItem = styled.div`
  border-bottom: 1px solid #ccc;

  &:last-child {
    border-bottom: none;
  }
`;

export const AccordionHeader = styled.div`
  padding: 10px;
  cursor: pointer;
  display: flex;
  align-items: center;
  color: #007bff;
`;

export const ChevronIcon = styled.div`
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 4px 4px 0 4px;
  border-color: #000 transparent transparent transparent;
  margin-left: auto;
  transition: transform 0.2s;
  ${({ expanded }: { expanded: boolean }) =>
    expanded && 'transform: rotate(180deg);'}
`;

export const AccordionContent = styled.div`
  background: #f5f5f5;
  padding: 10px 10px 20px 10px;
  display: none;
  ${({ expanded }: { expanded: boolean }) => expanded && 'display: block;'}
`;

export const EmptyStateContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const EmptyStateContent = styled.div`
  background: #f5f5f5;
  padding: 20px;
  width: 20%;
  text-align: center;
  border-radius: 10px;
`;
