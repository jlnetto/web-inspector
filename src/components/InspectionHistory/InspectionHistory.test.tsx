import { render, screen } from '@testing-library/react';
import InspectionHistory from './InspectionHistory';

describe('InspectionHistory', () => {
  it('renders title', () => {
    const { asFragment } = render(<InspectionHistory />);
    const titleElement = screen.getByText(
      'Histórico de solicitações cadastradas'
    );
    
    expect(titleElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders empty state when there are no records', () => {
    const { asFragment } = render(<InspectionHistory />);
    const emptyStateElement = screen.getByText('No records');
    
    expect(emptyStateElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders accordion items when there are records', () => {
    const storedData = [
      { id: '1', keyword: 'Keyword 1' },
      { id: '2', keyword: 'Keyword 2' },
    ];
    localStorage.setItem('lastKeywords', JSON.stringify(storedData));
    const { asFragment } = render(<InspectionHistory />);
    const accordionHeaders = screen.getAllByText(/Keyword \d/);

    expect(accordionHeaders.length).toBe(storedData.length);
    expect(asFragment()).toMatchSnapshot();
  });

});
