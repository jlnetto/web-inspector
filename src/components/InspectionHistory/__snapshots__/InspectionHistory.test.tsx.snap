// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspectionHistory renders accordion items when there are records 1`] = `
<DocumentFragment>
  <div
    class="sc-bgqQcB gNCgfU"
  >
    <h2
      class="sc-gTRrQi dvtghf"
    >
      Histórico de solicitações cadastradas
    </h2>
    <div
      class="sc-ewnqHT"
    >
      <div
        class="sc-fFGjHI fQDidE"
      >
        <div
          class="sc-dicizt gDDddz"
        >
          Keyword 1
          <div
            class="sc-iVCKna eeWChi"
          />
        </div>
        <div
          class="sc-gxYJeL iWjepY"
        >
          <div
            class="sc-crozmw bteBZu"
            data-testid="loading"
          />
        </div>
      </div>
      <div
        class="sc-fFGjHI fQDidE"
      >
        <div
          class="sc-dicizt gDDddz"
        >
          Keyword 2
          <div
            class="sc-iVCKna eeWChi"
          />
        </div>
        <div
          class="sc-gxYJeL iWjepY"
        >
          <div
            class="sc-crozmw bteBZu"
            data-testid="loading"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InspectionHistory renders empty state when there are no records 1`] = `
<DocumentFragment>
  <div
    class="sc-bgqQcB gNCgfU"
  >
    <h2
      class="sc-gTRrQi dvtghf"
    >
      Histórico de solicitações cadastradas
    </h2>
    <div
      class="sc-kPTPQs cFSfPo"
    >
      <div
        class="sc-ispOId hlbZba"
      >
        No records
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InspectionHistory renders title 1`] = `
<DocumentFragment>
  <div
    class="sc-bgqQcB gNCgfU"
  >
    <h2
      class="sc-gTRrQi dvtghf"
    >
      Histórico de solicitações cadastradas
    </h2>
    <div
      class="sc-kPTPQs cFSfPo"
    >
      <div
        class="sc-ispOId hlbZba"
      >
        No records
      </div>
    </div>
  </div>
</DocumentFragment>
`;
